### PR TITLE
V0.6.0

### PR DESCRIPTION
**Pull request by oguzhanunlu**
- [x] Ensure Stream Enrich resolver uses cacheTtl of one (#137)
- [x] Update example user_data.sh (#148)
- [x] Expose nsqadmin UI (#144)
- [x] Fix typo in README (#160)
- [x] Add ami-factory credentials to .travis.yml (#143)
- [x] Install NTP (#167)
- [x] Add gitter badge (#157)
- [x] Extend copyright to 2018 (#164)
- [x] Assemble the apps using Docker Compose (#23)
- [x] Add support for external Postgres for Iglu Server (#154)
- [x] Add publication of image for Google Compute Engine (#36)
- [x] Publish mini to bigger instance types (#173)
- [x] Bump Stream Enrich to 0.18.0 (#174)

#### release process
- [x] blog post (under review)
- [ ] discourse post
- [x] rc & qa (note to self: never trust in local dev env!)
- [x] doc update (prep a branch to be merged)
- [x] Prepared for release

### CODE REVIEW

**Comment by oguzhanunlu** [comment]
Hey @BenFradet , just added as reviewer, I'll ping you once I'm about to finish implementation phase.

**Comment by oguzhanunlu** [comment]
Hey @BenFradet , I'm done with the implementation phase and would like you to review my code changes.

I merged 4 tickets into #23 , see #23's page for included issues.

Also, only blocker for QA is that there is an issue with Caddy configuration (every endpoint except /kibana not working as expected, probably due to version bump, looking into that). I hope this is not a blocker for code review.

In the meantime, I'll solve the issue with /kibana endpoint, write a QA journal, prepare blog post first draft and a separate wiki branch to be merged.

File: provisioning/resources/control-plane/main.go [hunk 1]
```diff
@@ -115,6 +116,52 @@ func uploadEnrichments(resp http.ResponseWriter, req *http.Request) {
...
+
+		f, err := os.OpenFile(config.Dirs.Config+"/"+handler.Filename, os.O_WRONLY|os.O_CREATE, 0666)
+		if err != nil {
+			http.Error(resp, err.Error(), 500)
+			return
+		}
+		defer f.Close()
+
+		f.Truncate(0)
+		f.Seek(0, 0)
```
Comment (BenFradet) [inline provisioning/resources/control-plane/main.go]: can we add a comment here on what this is doing

File: provisioning/resources/elasticsearch/config/log4j2.properties [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+appender.rolling.policies.type = Policies
+appender.rolling.policies.time.type = TimeBasedTriggeringPolicy
+appender.rolling.policies.time.interval = 1
+appender.rolling.policies.time.modulate = true
+
+appender.rolling.strategy.type = DefaultRolloverStrategy
+appender.rolling.strategy.action.type = Delete
+appender.rolling.strategy.action.basepath = ${sys:es.logs.base_path}
+appender.rolling.strategy.action.condition.type = IfLastModified
+appender.rolling.strategy.action.condition.age = 7D
```
Comment (BenFradet) [inline provisioning/resources/elasticsearch/config/log4j2.properties]: isn't that too long?

File: provisioning/resources/elasticsearch/mapping/good-mapping.json [hunk 3]
```diff
@@ -248,79 +240,79 @@
...
-                    "type": "string",
-                    "index": "not_analyzed"
+                    "type": "keyword",
+                    "index": true
                 },
                 "v_tracker": {
-                    "type": "string",
-                    "index": "not_analyzed"
+                    "type": "keyword",
+                    "index": true
```
Comment (BenFradet) [inline provisioning/resources/elasticsearch/mapping/good-mapping.json]: what are the implications of those changes? cc @jbeemster

File: provisioning/resources/ui/js/components/Overview.tsx [hunk 4]
```diff
@@ -49,13 +48,14 @@ export class Overview extends React.Component<{}, {}> {
...
-          <li>Snowplow Iglu Server 0.2.0</li>
-          <li>NSQ 1.0.0</li>
-          <li>Elasticsearch 1.7.5</li>
-          <li>Kibana 4.0.1</li>
+          <li>Snowplow Elasticsearch Loader 0.10.1</li>
+          <li>Snowplow Iglu Server 0.3.0</li>
+          <li>Postgres 9.5</li>
+          <li>NSQ v1.0.0-compat</li>
+          <li>Elasticsearch 5.6.10</li>
+          <li>Kibana 5.6.10</li>
```
Comment (BenFradet) [inline provisioning/resources/ui/js/components/Overview.tsx]: aren't we better of with a 6.x version? cc @jbeemster

File: provisioning/roles/base/tasks/main.yml [hunk 5]
```diff
@@ -15,8 +15,8 @@
   lineinfile:
     dest="/etc/resolvconf/resolv.conf.d/base"
     line="nameserver 8.8.8.8"
-  sudo: yes
+  become: yes
 
 - name: Update nameserver list
   command: "resolvconf -u"
-  sudo: yes
+  become: yes
```
Comment (BenFradet) [inline provisioning/roles/base/tasks/main.yml]: what are the implications of those changes?

File: provisioning/roles/docker/files/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,171 @@
+version: "3"
+
+services:
+  elasticsearch:
+    image: elasticsearch:5.6.10
+    container_name: elasticsearch
+    restart: always
+    environment:
+      - "bootstrap.memory_lock=true"
```
Comment (BenFradet) [inline provisioning/roles/docker/files/docker-compose.yml]: what does this do? can we add a comment

File: provisioning/roles/docker/files/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+      - "ES_JAVA_OPTS=-Xms4g -Xmx4g"
```
Comment (BenFradet) [inline provisioning/roles/docker/files/docker-compose.yml]: should we be starting with 4g? cc @jbeemster

File: provisioning/roles/docker/files/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    depends_on:
+      - elasticsearch
+    volumes:
+      - /home/ubuntu/snowplow/configs:/snowplow/config
+    logging:
+      options:
+        max-size: "1M"
+        max-file: "10"
+    environment:
+      - "SP_JAVA_OPTS=-Xms512m -Xmx512m"
```
Comment (BenFradet) [inline provisioning/roles/docker/files/docker-compose.yml]: I wouldn't specify `Xms` for any of them

File: provisioning/roles/docker/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  shell: chmod +x /usr/local/bin/docker-compose
+
+- name: Copy docker-compose.yml
+  copy: src={{ item.src }} dest={{ item.dest }} owner=ubuntu group=ubuntu mode=0644
+  with_items:
+    - { src: '../files/docker-compose.yml', dest: '/home/ubuntu/snowplow/' }
+
+- name: Ensure file permissions
+  become: yes
+  shell: chown ubuntu:ubuntu -R /home/ubuntu/snowplow && chmod 755 -R /home/ubuntu/snowplow
```
Comment (BenFradet) [inline provisioning/roles/docker/tasks/main.yml]: why is this necessary, can we add a comment?

File: provisioning/roles/sp_mini_7_setup_init/tasks/main.yml [hunk 10]
```diff
@@ -9,29 +9,35 @@
...
-    - nsqd_init
-    - nsqlookupd_init
+
+- name: Start UI and Control Plane
+  become: yes
+  shell: service snowplow_mini_control_plane_api start && service caddy_init start
+
+- name: Increase mmap count to recommended 262144 for Elasticsearch
+  become: yes
+  shell: echo "vm.max_map_count=262144" >> /etc/sysctl.conf && service procps start
```
Comment (BenFradet) [inline provisioning/roles/sp_mini_7_setup_init/tasks/main.yml]: this isn't in the container why would it affect es?

File: provisioning/resources/configs/Caddyfile [hunk 11]
```diff
@@ -5,7 +5,6 @@
     /kibana
     /elasticsearch
     /control-plane
-    /_plugin
```
[the remark this answers is not in the document]
Comment (oguzhanunlu) [reply provisioning/resources/configs/Caddyfile]: Elasticsearch has removed site plugins as of 5.x,  but head plugin can be used as a [browser extension ](https://chrome.google.com/webstore/detail/elasticsearch-head/ffmkiejjmecolpfloofpjologoblkegm/) which works without a problem.

File: provisioning/resources/control-plane/main.go [hunk 12]
```diff
@@ -115,6 +116,52 @@ func uploadEnrichments(resp http.ResponseWriter, req *http.Request) {
...
+		// with the remainder stored on disk in temporary files
+		var maxMemory int64 = 32 << 20
+		err := req.ParseMultipartForm(maxMemory)
+
+		if err != nil {
+			http.Error(resp, err.Error(), 500)
+			return
+		}
+
+		file, handler, err := req.FormFile("igluserverjson")
```
[the remark this answers is not in the document]
Comment (oguzhanunlu) [reply provisioning/resources/control-plane/main.go]: Right, I'll correct.

File: provisioning/resources/elasticsearch/mapping/good-mapping.json (hunk 3 above)
Comment (oguzhanunlu) [reply provisioning/resources/elasticsearch/mapping/good-mapping.json]: `"string"` field is [unsupported](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/string.html) in favor of `text` and `keyword`. I followed Elastic's official [recommendations](https://www.elastic.co/blog/strings-are-dead-long-live-strings).

File: provisioning/roles/base/tasks/main.yml (hunk 5 above)
Comment (oguzhanunlu) [reply provisioning/roles/base/tasks/main.yml]: `sudo` is mostly used prior to Ansible 1.9 and [deprecated](https://docs.ansible.com/ansible/latest/user_guide/become.html#for-those-from-pre-1-9-sudo-and-su-still-work). We use Ansible 2.5 at the moment. Also, `googlecompute` builder of Packer doesn't recognize and ignore `sudo` which results in failed Ansible tasks.

File: provisioning/roles/docker/files/docker-compose.yml (hunk 6 above)
Comment (oguzhanunlu) [reply provisioning/roles/docker/files/docker-compose.yml]: It is explained [here](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/setup-configuration-memory.html#mlockall). I'll add a short comment.

File: provisioning/roles/docker/tasks/main.yml (hunk 9 above)
Comment (oguzhanunlu) [reply provisioning/roles/docker/tasks/main.yml]: I forgot to delete it, I used that for test purposes. 4th role already does that.

File: provisioning/resources/ui/js/components/Overview.tsx (hunk 4 above)
Comment (jbeemster) [reply provisioning/resources/ui/js/components/Overview.tsx]: Might as well yes.  Does the latest ES Sink support that?

File: provisioning/roles/sp_mini_7_setup_init/tasks/main.yml (hunk 10 above)
Comment (oguzhanunlu) [reply provisioning/roles/sp_mini_7_setup_init/tasks/main.yml]: I think it is because containers share the same kernel as host OS.

File: provisioning/resources/control-plane/main.go [hunk 13]
```diff
@@ -115,6 +116,52 @@ func uploadEnrichments(resp http.ResponseWriter, req *http.Request) {
...
+		if err != nil {
+			http.Error(resp, err.Error(), 500)
+			return
+		}
+		defer file.Close()
+
+		fileContentBytes, err := ioutil.ReadAll(file)
+		fileContent := string(fileContentBytes)
+
+		f, err := os.OpenFile(config.Dirs.Config+"/"+handler.Filename, os.O_WRONLY|os.O_CREATE, 0666)
```
[the remark this answers is not in the document]
Comment (oguzhanunlu) [reply provisioning/resources/control-plane/main.go]: No issues, I tested and works without a problem.

File: provisioning/resources/ui/js/components/Overview.tsx (hunk 4 above)
Comment (BenFradet) [reply provisioning/resources/ui/js/components/Overview.tsx]: http is fine :+1:

File: provisioning/roles/docker/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    port: 9200
+    delay: 10
+    sleep: 5
+    connect_timeout: 60
+
+# Kibana 5.x has issues updating index patterns
+# rebooting kibana and retrying works out
+- name: Likely to fail attempt to create ES indexes & Kibana index patterns & NSQ topics
+  become: yes
+  shell: sh {{init_dir}}/create.sh
```
Comment (BenFradet) [inline provisioning/roles/docker/tasks/main.yml]: what's the reasoning for this

File: provisioning/roles/docker/files/docker-compose.yml (hunk 8 above)
Comment (BenFradet) [reply provisioning/roles/docker/files/docker-compose.yml]: why was this removed altogether, I would still specify `Xmx`

File: provisioning/resources/ui/js/components/Overview.tsx (hunk 4 above)
Comment (BenFradet) [reply provisioning/resources/ui/js/components/Overview.tsx]: this wasn't done

File: provisioning/roles/docker/files/docker-compose.yml (hunk 7 above)
Comment (BenFradet) [reply provisioning/roles/docker/files/docker-compose.yml]: what's the reasoning behind those settings?

File: provisioning/resources/elasticsearch/config/log4j2.properties (hunk 2 above)
Comment (BenFradet) [reply provisioning/resources/elasticsearch/config/log4j2.properties]: this wasn't acted on

File: provisioning/roles/sp_mini_7_setup_init/tasks/main.yml [hunk 15]
```diff
@@ -9,29 +9,35 @@
...
+    name: "Add cronjob to restart containers at system reboot"
+    special_time: reboot
+    job: /usr/local/bin/docker-compose -f /home/ubuntu/snowplow/docker-compose.yml restart && sleep 30
+
+- cron:
+    name: "Add cronjob to crop ES good index's docs older than a week"
+    special_time: weekly
+    job: /usr/bin/curl -s -X POST http://localhost:9200/good/_delete_by_query -d '{ "query" :{ "range" :{ "collector_tstamp" :{ "lt" :"now-1w/d" } } } }' > /dev/null 2>&1
+
+- cron:
```
Comment (BenFradet) [inline provisioning/roles/sp_mini_7_setup_init/tasks/main.yml]: can we have something more specific as far as names are concerned instead of 3 times `cron`?

**Comment by chuwy** [comment]
Am I right that we did not bump Iglu Server to 0.3.0 in this (or previous) release? If we didn't - why? It looked like a quite serious upgrade. If we did - it probably does not work as expected due https://github.com/snowplow/iglu/issues/378

File: provisioning/roles/docker/files/docker-compose.yml (hunk 7 above)
Comment (BenFradet) [reply provisioning/roles/docker/files/docker-compose.yml]: we need to take the size of the machine into account now c.f. #173, Xms and Xmx should indeed be equal, c.f. https://www.elastic.co/guide/en/elasticsearch/reference/6.2/heap-size.html

File: provisioning/resources/elasticsearch/config/log4j2.properties (hunk 2 above)
Comment (oguzhanunlu) [reply provisioning/resources/elasticsearch/config/log4j2.properties]: How many days of log should be kept @BenFradet ? Is a week unnecessarily long here?

Comment (BenFradet) [reply provisioning/resources/elasticsearch/config/log4j2.properties]: 3 days seems long enough to me, what do you reckon @jbeemster ?

Comment (jbeemster) [reply provisioning/resources/elasticsearch/config/log4j2.properties]: SGTM!

File: provisioning/roles/docker/files/docker-compose.yml (hunk 7 above)
Comment (oguzhanunlu) [reply provisioning/roles/docker/files/docker-compose.yml]: Let's decide and I'll act accordingly regarding heap size. From the page @BenFradet  referenced, it is indicated as following;

```
Set Xmx to no more than 50% of your physical RAM, to ensure that there is enough
physical RAM left for kernel file system caches
```

My only reasoning is to make ES able to stay up under high load, as much as possible. I remember us talking about publishing to `t2.large` as of this release and it has 8GB ram, hence 4GB allocation to ES.

Should it be less or is it fine, what do you think? @BenFradet @jbeemster

File: provisioning/roles/docker/tasks/main.yml (hunk 14 above)
Comment (oguzhanunlu) [reply provisioning/roles/docker/tasks/main.yml]: Kibana 5.x had problems with creating index patterns (they still don't have a stable API documentation, unfortunately) and I guaranteed index pattern creation using a dirty way. Anyway, 6.x doesn't have aforementioned issue and I removed this part.

Comment (BenFradet) [reply provisioning/roles/docker/tasks/main.yml]: nice :+1:

File: provisioning/roles/docker/files/docker-compose.yml (hunk 8 above)
Comment (oguzhanunlu) [reply provisioning/roles/docker/files/docker-compose.yml]: Specified only Xmx now.

File: provisioning/resources/ui/js/components/Overview.tsx (hunk 4 above)
Comment (oguzhanunlu) [reply provisioning/resources/ui/js/components/Overview.tsx]: Done now

File: provisioning/resources/elasticsearch/config/log4j2.properties (hunk 2 above)
Comment (oguzhanunlu) [reply provisioning/resources/elasticsearch/config/log4j2.properties]: Done

File: provisioning/roles/docker/files/docker-compose.yml (hunk 7 above)
Comment (BenFradet) [reply provisioning/roles/docker/files/docker-compose.yml]: I think 50% is good so that would be:

type | ES | loader | enrich | collector
| --- | --- | --- | --- | ---
t2.large | 4G | 0.5G | 0.5G | 0.5G
t2.xlarge | 8G | 1.5G | 1.5G | 1.5G
t2.2xlarge |16G | 3G | 3G | 3G

**Comment by oguzhanunlu** [comment]
Hey @chuwy , @BenFradet has been through his QA and he thinks it is good to go. Will you also QA or should we release tomorrow?

**Comment by chuwy** [comment]
Good to go from me.
